starlette/middleware/base.py

```diff
@@ -42,18 +42,41 @@ async def coro() -> None:
         task = loop.create_task(coro())
         message = await queue.get()
         if message is None:
+            queue.task_done()
             task.result()
             raise RuntimeError("No response returned.")
         assert message["type"] == "http.response.start"
 
         async def body_stream() -> typing.AsyncGenerator[bytes, None]:
-            while True:
+            def streaming_predicate(
+                msg: typing.Optional[dict], more_body: bool = True
+            ) -> bool:
+                return (
+                    msg is not None
+                    and msg["type"] == "http.response.body"
+                    and "more_body" in msg
+                    and msg["more_body"] is more_body
+                )
+
+            # In non-streaming responses, there will be one message to emit
+            message = await queue.get()
+            queue.task_done()
+            assert message["type"] == "http.response.body"
```

Comment:
There is a possibility that an exception is thrown when the body _starts_ in which case this message will also be `None`. I need to handle that case, I believe. I don't have a test case for it yet, so I haven't written the code.